src/util/expr_iterator.h

```diff
@@ -0,0 +1,228 @@
+  }
+  bool operator!=(const depth_iterator_t &other) const
+  {
+    return !(*this==other);
+  }
+  depth_iterator_t &operator++()
+  {
+    PRECONDITION(!m_stack.empty());
+    while(true)
+    {
+      if(m_stack.back().it==m_stack.back().end)
+      {
+        m_stack.pop_back();
+        if(m_stack.empty())
+        {
+          break;
+        }
+      }
+      else if(this->downcast().push_expr(*m_stack.back().it))
+      {
+        break;
+      }
+      m_stack.back().it++;
+    }
+    return this->downcast();
+  }
+  const exprt &operator*() const
+  {
+    return m_stack.back().expr.get();
+  }
+  const exprt *operator->() const
+  {
+    return &**this;
+  }
+protected:
+  // Create end iterator
+  depth_iterator_baset() { }
+  // Create begin iterator, starting at the supplied node
+  explicit depth_iterator_baset(const exprt &root)
+  {
+    this->downcast().push_expr(root);
+  }
+  typedef typename std::vector<exprt>::const_iterator operands_iteratort;
+
+  struct expr_statet final
+  {
+    expr_statet(
+      const exprt &expr,
+      operands_iteratort it,
+      operands_iteratort end):
+      expr(expr), it(it), end(end) { }
+    expr_statet(const expr_statet &other):
+      expr(other.expr), it(other.it), end(other.end) { }
+    std::reference_wrapper<const exprt> expr;
+    // Cache to avoid unnecessary mutable operands() calls
+    std::vector<exprt> *operands=nullptr;
+    operands_iteratort it;
+    const operands_iteratort end;
+    bool operator==(const expr_statet &other) const
+    {
+      return it==other.it;
+    }
+  };
+  // Obtain non-const exprt reference. Breaks the tree as a side
+  // effect. To be called only if a non-const exprt was passed into
+  // the constructor Do not call on the end() iterator
+  exprt &mutate()
+  {
+    PRECONDITION(!m_stack.empty());
+    auto expr=std::ref(const_cast<exprt&>(m_stack.front().expr.get()));
+    auto old_stack=std::move(m_stack);
+    try
+    {
+      m_stack.reserve(old_stack.size());
+      for(const auto& state : old_stack)
+      {
+        auto &operands=state.operands
+          ? *state.operands
+          : expr.get().operands();
+        const auto i=operands.size()-(state.end-state.it);
+        const auto it=operands.begin()+i;
+        const auto end=operands.end();
+        m_stack.emplace_back(expr.get(), it, end);
+        m_stack.back().operands = &operands;
+        if(!(state==old_stack.back()))
+        {
+          expr=std::ref(*it);
+        }
+      }
+      return expr.get();
+    }
+    catch(...)
+    {
+      m_stack=std::move(old_stack);
+      throw;
+    }
+  }
+  // returns true if element was successfully pushed onto the stack
+  bool push_expr(const exprt &expr)
+  {
+    m_stack.emplace_back(expr, expr.operands().begin(), expr.operands().end());
+    return true;
+  }
+  std::vector<expr_statet> m_stack;
+  // Segfaults if stored in unique_depth_iterato_baset
+  std::set<exprt> m_traversed;
```

Comment:
That's a problem of the order of initialisation. You can fix this by doing:
```
explicit unique_depth_iterator_baset(const exprt &expr):
    depth_iterator_baset<depth_iterator_t>() { push_expr(expr); }
```
below.